File: dev/tests/schema-l1/redis-failure.php
<?php

require (getenv('MAGENTO_ROOT') ?: dirname(__DIR__, 5)).'/vendor/autoload.php';
use GraphCommerce\FastBootCache\Model\Schema\{Remote,Local};

[$script,$port,$dir,$mode] = $argv;
$options = ['host' => '127.0.0.1','port' => (int)$port,'database' => 0,'key' => 'failure-fixture','password' => 'audit-fixture-only','timeout' => 0.15,'read_timeout' => 0.15];
if ($mode === 'persistent-acl') {
    $options['username'] = 'fastboot-fixture';
    $options['password'] = 'acl-fixture-only';
    $options['database'] = 3;
}
$r = new Remote($options);
$l = new Local($r, $dir, 0);
if (in_array($mode, ['persistent', 'persistent-acl'], true)) {
    $r->save('{"ok":1}', 'schema', [], 60);
    $l->load();
    echo "READY\n";
    flush();
    while (($action = trim((string)fgets(STDIN))) !== 'quit') {
        if ($action === 'timeout') {
            try {
                $l->load();
                throw new RuntimeException('Expected timeout');
            } catch (RedisException|CredisException) {
            }
        } elseif ($action === 'recovered') {
            if ($l->load() !== ['ok' => 1]) {
                throw new RuntimeException('Same-client recovery failed');
            }
        } else {
            throw new RuntimeException('Unknown action');
        }
        echo "PASS $action\n";
        flush();
    }
    exit;
}

switch ($mode) {
    case 'acl':
        $r->command('acl', ['SETUSER', 'fastboot-fixture', 'on', '>acl-fixture-only', '~*', '+@all']);
        $options['username'] = 'fastboot-fixture';
        $options['password'] = 'acl-fixture-only';
        $options['database'] = 3;
        $acl = new Remote($options);
        $acl->save('{"acl":true}', 'schema', [], 60);
        if ($acl->load('schema') !== '{"acl":true}' || $r->load('schema') !== '{"ok":1}') {
            throw new RuntimeException('ACL authentication or database isolation failed');
        }
        break;
    case 'seed':$r->save('{"ok":1}', 'schema', [], 60);
        if ($l->load() !== ['ok' => 1]) {
            throw new RuntimeException('Seed failure');
        }break;
    case 'bad-auth':$options['password'] = 'wrong';
        try {
            (new Remote($options))->metadata();
            throw new RuntimeException('Wrong credentials accepted');
        } catch (RedisException|CredisException) {
        }break;
    case 'failed':$start = microtime(true);
        try {
            $l->load();
            throw new RuntimeException('Stale value accepted');
        } catch (RedisException|CredisException) {
        }if (microtime(true) - $start > 2) {
            throw new RuntimeException('Unbounded timeout');
        }break;
    case 'grace':if ((new Local($r, $dir, 60))->load() !== ['ok' => 1]) {
        throw new RuntimeException('Expected bounded grace');
    }break;
    case 'empty':if ($l->load() !== null) {
        throw new RuntimeException('Restart resurrected L1');
    }break;
    default:throw new RuntimeException('Bad mode');
}
echo "PASS $mode\n";

File: dev/tests/schema-l1/redis-failure.py
import os
import subprocess,pathlib,secrets,json,time,socket
D=pathlib.Path(__file__).resolve().parent
name='fba-schema-chaos-'+secrets.token_hex(5)
def docker(*args):return subprocess.check_output(['docker',*args],text=True).strip()
def check(mode):subprocess.run(['php',str(D/'redis-failure.php'),port,str(pathlib.Path(os.environ.get('MAGENTO_ROOT',D.parents[4]))/'var'/name),mode],check=True,timeout=5)
try:
 docker('run','-d','--name',name,'-p','127.0.0.1::6379','valkey/valkey:8-alpine','valkey-server','--save','','--appendonly','no','--requirepass','audit-fixture-only')
 port=docker('port',name,'6379/tcp').rsplit(':',1)[1]
 time.sleep(.5)
 check('seed');check('bad-auth');check('acl')
 live=subprocess.Popen(['php',str(D/'redis-failure.php'),port,str(pathlib.Path(os.environ.get('MAGENTO_ROOT',D.parents[4]))/'var'/name/'persistent'),'persistent-acl'],stdin=subprocess.PIPE,stdout=subprocess.PIPE,stderr=subprocess.PIPE,text=True)
 try:
  assert live.stdout.readline().strip()=='READY'
  docker('pause',name)
  try:
   live.stdin.write('timeout\n');live.stdin.flush();assert live.stdout.readline().strip()=='PASS timeout'
  finally:docker('unpause',name)
  live.stdin.write('recovered\n');live.stdin.flush();assert live.stdout.readline().strip()=='PASS recovered'
  print('PASS same PHP process and Remote object recover after real read timeout')
 finally:
  live.communicate('quit\n',timeout=5)
  if live.returncode:raise RuntimeError('Persistent client failed')
 docker('pause',name)
 try:check('failed');check('grace')
 finally:docker('unpause',name)
 docker('restart',name)
 port=docker('port',name,'6379/tcp').rsplit(':',1)[1]
 for _ in range(50):
  try:
   with socket.create_connection(('127.0.0.1',int(port)),timeout=.1):break
  except OSError:time.sleep(.1)
 check('empty');check('seed')
 docker('stop',name);check('failed')
 print('PASS actual Redis pause/timeout, bad authentication, empty restart, recovery and connection refusal')
finally:
 subprocess.run(['docker','rm','-f',name],stdout=subprocess.DEVNULL,stderr=subprocess.DEVNULL)
